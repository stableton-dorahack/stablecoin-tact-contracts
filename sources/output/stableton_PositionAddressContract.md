# TACT Compilation Report
Contract: PositionAddressContract
BOC Size: 572 bytes

# Types
Total Types: 19

## StateInit
TLB: `_ code:^cell data:^cell = StateInit`
Signature: `StateInit{code:^cell,data:^cell}`

## Context
TLB: `_ bounced:bool sender:address value:int257 raw:^slice = Context`
Signature: `Context{bounced:bool,sender:address,value:int257,raw:^slice}`

## SendParameters
TLB: `_ bounce:bool to:address value:int257 mode:int257 body:Maybe ^cell code:Maybe ^cell data:Maybe ^cell = SendParameters`
Signature: `SendParameters{bounce:bool,to:address,value:int257,mode:int257,body:Maybe ^cell,code:Maybe ^cell,data:Maybe ^cell}`

## Deploy
TLB: `deploy#946a98b6 queryId:uint64 = Deploy`
Signature: `Deploy{queryId:uint64}`

## DeployOk
TLB: `deploy_ok#aff90f57 queryId:uint64 = DeployOk`
Signature: `DeployOk{queryId:uint64}`

## PositionState
TLB: `_ collateral:coins debt:uint64 = PositionState`
Signature: `PositionState{collateral:coins,debt:uint64}`

## StablecoinBurnedMessage
TLB: `stablecoin_burned_message#d2be4c6c user:address amount:uint64 fees:uint64 = StablecoinBurnedMessage`
Signature: `StablecoinBurnedMessage{user:address,amount:uint64,fees:uint64}`

## DepositCollateralMessage
TLB: `deposit_collateral_message#11dfb138 user:address amount:coins = DepositCollateralMessage`
Signature: `DepositCollateralMessage{user:address,amount:coins}`

## WithdrawCollateralMessage
TLB: `withdraw_collateral_message#6829d55e user:address amount:coins debtRate:int257 = WithdrawCollateralMessage`
Signature: `WithdrawCollateralMessage{user:address,amount:coins,debtRate:int257}`

## WithdrawStablecoinMessage
TLB: `withdraw_stablecoin_message#f3554cd7 user:address amount:uint64 debtRate:uint64 = WithdrawStablecoinMessage`
Signature: `WithdrawStablecoinMessage{user:address,amount:uint64,debtRate:uint64}`

## RepayStablecoinMessage
TLB: `repay_stablecoin_message#6e81f446 user:address amount:uint64 debtRate:uint64 = RepayStablecoinMessage`
Signature: `RepayStablecoinMessage{user:address,amount:uint64,debtRate:uint64}`

## DoWithdrawCollateralMessage
TLB: `do_withdraw_collateral_message#c3cd33bf user:address amount:uint64 = DoWithdrawCollateralMessage`
Signature: `DoWithdrawCollateralMessage{user:address,amount:uint64}`

## IncreaseTotalStableMessage
TLB: `increase_total_stable_message#96096519 user:address amount:uint64 = IncreaseTotalStableMessage`
Signature: `IncreaseTotalStableMessage{user:address,amount:uint64}`

## DecreaseTotalStableMessage
TLB: `decrease_total_stable_message#ed81304a user:address amount:uint64 fees:uint64 = DecreaseTotalStableMessage`
Signature: `DecreaseTotalStableMessage{user:address,amount:uint64,fees:uint64}`

## MintFeesMessage
TLB: `mint_fees_message#9d42541d to:address amount:coins = MintFeesMessage`
Signature: `MintFeesMessage{to:address,amount:coins}`

## JettonData
TLB: `_ totalSupply:int257 mintable:bool owner:address content:Maybe ^cell walletCode:^cell = JettonData`
Signature: `JettonData{totalSupply:int257,mintable:bool,owner:address,content:Maybe ^cell,walletCode:^cell}`

## JettonWalletData
TLB: `_ balance:int257 owner:address master:address walletCode:^cell = JettonWalletData`
Signature: `JettonWalletData{balance:int257,owner:address,master:address,walletCode:^cell}`

## SetPositionIdMessage
TLB: `set_position_id_message#00b3cad5 positionId:uint8 = SetPositionIdMessage`
Signature: `SetPositionIdMessage{positionId:uint8}`

## SetPositionAddressMessage
TLB: `set_position_address_message#cf891f9e user:address = SetPositionAddressMessage`
Signature: `SetPositionAddressMessage{user:address}`

# Get Methods
Total Get Methods: 2

## getPositionId

## getPositionAddress

# Error Codes
2: Stack undeflow
3: Stack overflow
4: Integer overflow
5: Integer out of expected range
6: Invalid opcode
7: Type check error
8: Cell overflow
9: Cell underflow
10: Dictionary error
13: Out of gas error
32: Method ID not found
34: Action is invalid or not supported
37: Not enough TON
38: Not enough extra-currencies
128: Null reference exception
129: Invalid serialization prefix
130: Invalid incoming message
131: Constraints error
132: Access denied
133: Contract stopped
134: Invalid argument
135: Code of a contract was not found
136: Invalid address
4429: Invalid sender