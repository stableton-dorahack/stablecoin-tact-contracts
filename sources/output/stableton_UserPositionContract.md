# TACT Compilation Report
Contract: UserPositionContract
BOC Size: 1678 bytes

# Types
Total Types: 25

## StateInit
TLB: `_ code:^cell data:^cell = StateInit`
Signature: `StateInit{code:^cell,data:^cell}`

## Context
TLB: `_ bounced:bool sender:address value:int257 raw:^slice = Context`
Signature: `Context{bounced:bool,sender:address,value:int257,raw:^slice}`

## SendParameters
TLB: `_ bounce:bool to:address value:int257 mode:int257 body:Maybe ^cell code:Maybe ^cell data:Maybe ^cell = SendParameters`
Signature: `SendParameters{bounce:bool,to:address,value:int257,mode:int257,body:Maybe ^cell,code:Maybe ^cell,data:Maybe ^cell}`

## Deploy
TLB: `deploy#946a98b6 queryId:uint64 = Deploy`
Signature: `Deploy{queryId:uint64}`

## DeployOk
TLB: `deploy_ok#aff90f57 queryId:uint64 = DeployOk`
Signature: `DeployOk{queryId:uint64}`

## ChangeOwner
TLB: `change_owner#0f474d03 newOwner:address = ChangeOwner`
Signature: `ChangeOwner{newOwner:address}`

## DepositCollateralMessage
TLB: `deposit_collateral_message#11dfb138 user:address amount:coins = DepositCollateralMessage`
Signature: `DepositCollateralMessage{user:address,amount:coins}`

## WithdrawCollateralMessage
TLB: `withdraw_collateral_message#6829d55e user:address amount:coins debtRate:int257 = WithdrawCollateralMessage`
Signature: `WithdrawCollateralMessage{user:address,amount:coins,debtRate:int257}`

## WithdrawStablecoinMessage
TLB: `withdraw_stablecoin_message#f3554cd7 user:address amount:uint64 debtRate:uint64 = WithdrawStablecoinMessage`
Signature: `WithdrawStablecoinMessage{user:address,amount:uint64,debtRate:uint64}`

## RepayStablecoinMessage
TLB: `repay_stablecoin_message#6e81f446 user:address amount:uint64 debtRate:uint64 = RepayStablecoinMessage`
Signature: `RepayStablecoinMessage{user:address,amount:uint64,debtRate:uint64}`

## DoWithdrawCollateralMessage
TLB: `do_withdraw_collateral_message#c3cd33bf user:address amount:uint64 = DoWithdrawCollateralMessage`
Signature: `DoWithdrawCollateralMessage{user:address,amount:uint64}`

## IncreaseTotalStableMessage
TLB: `increase_total_stable_message#96096519 user:address amount:uint64 = IncreaseTotalStableMessage`
Signature: `IncreaseTotalStableMessage{user:address,amount:uint64}`

## DecreaseTotalStableMessage
TLB: `decrease_total_stable_message#ed81304a user:address amount:uint64 fees:uint64 = DecreaseTotalStableMessage`
Signature: `DecreaseTotalStableMessage{user:address,amount:uint64,fees:uint64}`

## MintFeesMessage
TLB: `mint_fees_message#9d42541d to:address amount:coins = MintFeesMessage`
Signature: `MintFeesMessage{to:address,amount:coins}`

## JettonData
TLB: `_ totalSupply:int257 mintable:bool owner:address content:Maybe ^cell walletCode:^cell = JettonData`
Signature: `JettonData{totalSupply:int257,mintable:bool,owner:address,content:Maybe ^cell,walletCode:^cell}`

## JettonWalletData
TLB: `_ balance:int257 owner:address master:address walletCode:^cell = JettonWalletData`
Signature: `JettonWalletData{balance:int257,owner:address,master:address,walletCode:^cell}`

## SetUserPositionDependecyMessage
TLB: `set_user_position_dependecy_message#f037fcfb stablecoinMasterAddress:address positionsManagerAddress:address = SetUserPositionDependecyMessage`
Signature: `SetUserPositionDependecyMessage{stablecoinMasterAddress:address,positionsManagerAddress:address}`

## SetPositionIdMessage
TLB: `set_position_id_message#4a35f9d0 user:address positionId:int257 = SetPositionIdMessage`
Signature: `SetPositionIdMessage{user:address,positionId:int257}`

## NewPositionIdMessage
TLB: `new_position_id_message#0d74c595 user:address = NewPositionIdMessage`
Signature: `NewPositionIdMessage{user:address}`

## SetPositionAddressMessage
TLB: `set_position_address_message#5da6de96 user:address position:address = SetPositionAddressMessage`
Signature: `SetPositionAddressMessage{user:address,position:address}`

## MintMessage
TLB: `mint_message#903559b2 user:address amount:coins = MintMessage`
Signature: `MintMessage{user:address,amount:coins}`

## BurnMessage
TLB: `burn_message#b625719b user:address amount:coins fees:coins = BurnMessage`
Signature: `BurnMessage{user:address,amount:coins,fees:coins}`

## RepayBurnNotification
TLB: `repay_burn_notification#563c6d94 user:address amount:coins = RepayBurnNotification`
Signature: `RepayBurnNotification{user:address,amount:coins}`

## StablecoinBurnedMessage
TLB: `stablecoin_burned_message#d2be4c6c user:address amount:uint64 fees:uint64 = StablecoinBurnedMessage`
Signature: `StablecoinBurnedMessage{user:address,amount:uint64,fees:uint64}`

## PositionState
TLB: `_ collateral:coins debt:uint64 = PositionState`
Signature: `PositionState{collateral:coins,debt:uint64}`

# Get Methods
Total Get Methods: 4

## getPositionUser

## getPositionId

## getPositionState

## owner

# Error Codes
2: Stack undeflow
3: Stack overflow
4: Integer overflow
5: Integer out of expected range
6: Invalid opcode
7: Type check error
8: Cell overflow
9: Cell underflow
10: Dictionary error
13: Out of gas error
32: Method ID not found
34: Action is invalid or not supported
37: Not enough TON
38: Not enough extra-currencies
128: Null reference exception
129: Invalid serialization prefix
130: Invalid incoming message
131: Constraints error
132: Access denied
133: Contract stopped
134: Invalid argument
135: Code of a contract was not found
136: Invalid address
15032: not from stablecoin master
22230: Already set
31797: debt less than repay amount
53160: not from positions manager
61910: not from positionsManager
63577: withdrawal amount more than position has