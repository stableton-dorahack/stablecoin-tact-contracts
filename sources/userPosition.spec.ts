import { Address, Sender, toNano } from "ton";
import { ContractSystem, Treasure } from "ton-emulator";
import { PositionAddressContract } from "./output/stableton_PositionAddressContract";
import { UserPositionContract } from "./output/stableton_UserPositionContract";

describe.only("UserPositionContract", () => {
    let system: ContractSystem;
    let owner: Treasure;
    let user: Treasure;

    beforeAll(async () => {
        system = await ContractSystem.create();
        owner = system.treasure("owner");
        user = system.treasure("user");
    });

    it("should deploy correctly", async () => {
        let userPositionContract = system.open(await UserPositionContract.fromInit(user.address));

        let track = system.track(userPositionContract.address);

        await userPositionContract.send(owner, { value: toNano(1) }, { $$type: "Deploy", queryId: 0n });

        await system.run();

        expect(track.collect()).toMatchSnapshot();

        const getPositionUser = (await userPositionContract.getGetPositionUser()).toString();

        expect(getPositionUser).toEqual(user.address.toString());
    });

    // it("should set position address if called by userPosition contract", async () => {
    //     let system = await ContractSystem.create();
    //     let owner: Treasure = system.treasure("owner");
    //     let user = system.treasure("user");
    //     let position = system.treasure("position");
    //     console.log("position address", position.address);

    //     let id = 1n;
    //     let positionAddressContract = system.open(await PositionAddressContract.fromInit(id));

    //     let userPositionContract = system.open(await UserPositionContract.fromAddress(position.address));

    //     let track = system.track(positionAddressContract.address);

    //     await positionAddressContract.send(owner, { value: toNano(1) }, { $$type: "Deploy", queryId: 0n });
    //     // await system.run();
    //     await positionAddressContract.send(
    //         position,
    //         { value: toNano(1) },
    //         { $$type: "SetPositionAddressMessage", user: user.address }
    //     );

    //     await system.run();
    //     expect(track.collect()).toMatchSnapshot();

    //     const setAddress = (await positionAddressContract.getGetPositionAddress()).toString();

    //     console.log("address set", setAddress);

    //     expect(setAddress).toEqual(position.address.toString());
    //     expect(await positionAddressContract.getGetPositionId()).toEqual(id);
    // });
});
