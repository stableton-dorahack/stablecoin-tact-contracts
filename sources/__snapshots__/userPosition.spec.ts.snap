// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PositionsManagerContract on addCollateral from positionsManager 1`] = `
[
  {
    "$seq": 5,
    "events": [
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{DDED1FE8800784F79BB9BDDE39E2B05A2820A591FBCD92F3ACC083685CD65E37ED00489A4B2A2E90EDD000000000020000000200000002000000020000000200000002935F1B401_}",
            "type": "cell",
          },
          "bounce": true,
          "from": "kQCmw0-Nvnap6F8s-77yXEYbduFtd_bSsmDZu6eYT4Oy-Aqh",
          "to": "kQDhiRIXBxafSMMm6UXqSB2b7t0IWaAHrHiNHLJd5ORy-0BR",
          "type": "internal",
          "value": 1000000000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 11744n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "type": "empty",
            },
            "bounce": false,
            "from": "kQDhiRIXBxafSMMm6UXqSB2b7t0IWaAHrHiNHLJd5ORy-0BR",
            "to": "kQA8J7zdze7xzxWC0UEFLI_ebJedZgQbQuay8b9oAkTSWQkO",
            "type": "internal",
            "value": 987256000n,
          },
        ],
      },
    ],
  },
]
`;

exports[`PositionsManagerContract on repayStablecoin from positionsManager with healthy position 1`] = `
[
  {
    "$seq": 11,
    "events": [
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{BEF31E07800784F79BB9BDDE39E2B05A2820A591FBCD92F3ACC083685CD65E37ED00489A4B200000000000000C8000000020000000200000002000000020000000200000002935F1B401_}",
            "type": "cell",
          },
          "bounce": true,
          "from": "kQCmw0-Nvnap6F8s-77yXEYbduFtd_bSsmDZu6eYT4Oy-Aqh",
          "to": "kQDhiRIXBxafSMMm6UXqSB2b7t0IWaAHrHiNHLJd5ORy-0BR",
          "type": "internal",
          "value": 1000000000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 13471n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "cell": "x{B065C3AC800784F79BB9BDDE39E2B05A2820A591FBCD92F3ACC083685CD65E37ED00489A4B22C8203_}",
              "type": "cell",
            },
            "bounce": true,
            "from": "kQDhiRIXBxafSMMm6UXqSB2b7t0IWaAHrHiNHLJd5ORy-0BR",
            "to": "kQCDr0Gm-nivCD9KWPb6JkiDulGZ8eNb2Xm_HNi39oyNbBwJ",
            "type": "internal",
            "value": 498577000n,
          },
        ],
      },
    ],
  },
  {
    "$seq": 12,
    "events": [
      {
        "$type": "received-bounced",
        "message": {
          "body": {
            "type": "empty",
          },
          "bounce": false,
          "from": "kQCDr0Gm-nivCD9KWPb6JkiDulGZ8eNb2Xm_HNi39oyNbBwJ",
          "to": "kQDhiRIXBxafSMMm6UXqSB2b7t0IWaAHrHiNHLJd5ORy-0BR",
          "type": "internal",
          "value": 497577000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 7461n,
      },
    ],
  },
]
`;

exports[`PositionsManagerContract on repayStablecoin from positionsManager with healthy position revert if debt less than amount 1`] = `
[
  {
    "$seq": 9,
    "events": [
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{BEF31E07800784F79BB9BDDE39E2B05A2820A591FBCD92F3ACC083685CD65E37ED00489A4B2000000000000012C000000020000000200000002000000020000000200000002935F1B401_}",
            "type": "cell",
          },
          "bounce": true,
          "from": "kQCmw0-Nvnap6F8s-77yXEYbduFtd_bSsmDZu6eYT4Oy-Aqh",
          "to": "kQDhiRIXBxafSMMm6UXqSB2b7t0IWaAHrHiNHLJd5ORy-0BR",
          "type": "internal",
          "value": 1000000000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 13471n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "cell": "x{B065C3AC800784F79BB9BDDE39E2B05A2820A591FBCD92F3ACC083685CD65E37ED00489A4B232C203_}",
              "type": "cell",
            },
            "bounce": true,
            "from": "kQDhiRIXBxafSMMm6UXqSB2b7t0IWaAHrHiNHLJd5ORy-0BR",
            "to": "kQCDr0Gm-nivCD9KWPb6JkiDulGZ8eNb2Xm_HNi39oyNbBwJ",
            "type": "internal",
            "value": 498577000n,
          },
        ],
      },
    ],
  },
  {
    "$seq": 10,
    "events": [
      {
        "$type": "received-bounced",
        "message": {
          "body": {
            "type": "empty",
          },
          "bounce": false,
          "from": "kQCDr0Gm-nivCD9KWPb6JkiDulGZ8eNb2Xm_HNi39oyNbBwJ",
          "to": "kQDhiRIXBxafSMMm6UXqSB2b7t0IWaAHrHiNHLJd5ORy-0BR",
          "type": "internal",
          "value": 497577000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 7461n,
      },
    ],
  },
]
`;

exports[`PositionsManagerContract on setPositionId from positionsManager should set positionId and deploy positionAddress contract 1`] = `
[
  {
    "$seq": 2,
    "events": [
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{4A35F9D0800784F79BB9BDDE39E2B05A2820A591FBCD92F3ACC083685CD65E37ED00489A4B20000000000000000000000000000000000000000000000000000000000000017}",
            "type": "cell",
          },
          "bounce": true,
          "from": "kQCmw0-Nvnap6F8s-77yXEYbduFtd_bSsmDZu6eYT4Oy-Aqh",
          "to": "kQDhiRIXBxafSMMm6UXqSB2b7t0IWaAHrHiNHLJd5ORy-0BR",
          "type": "internal",
          "value": 1000000000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 20413n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "cell": "x{5DA6DE96800784F79BB9BDDE39E2B05A2820A591FBCD92F3ACC083685CD65E37ED00489A4B30038624485C1C5A7D230C9BA517A920766FBB742166801EB1E23472C9779391CBEE_}",
              "type": "cell",
            },
            "bounce": true,
            "from": "kQDhiRIXBxafSMMm6UXqSB2b7t0IWaAHrHiNHLJd5ORy-0BR",
            "to": "kQCe9rE3xPdst3hC14pLYI445Rop9zC6414J3Dar2QoLJmVd",
            "type": "internal",
            "value": 966560000n,
          },
        ],
      },
    ],
  },
]
`;

exports[`PositionsManagerContract on stablecoin burned  1`] = `
[
  {
    "$seq": 13,
    "events": [
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{EA319D9C800784F79BB9BDDE39E2B05A2820A591FBCD92F3ACC083685CD65E37ED00489A4B22C8203_}",
            "type": "cell",
          },
          "bounce": true,
          "from": "kQCDr0Gm-nivCD9KWPb6JkiDulGZ8eNb2Xm_HNi39oyNbBwJ",
          "to": "kQDhiRIXBxafSMMm6UXqSB2b7t0IWaAHrHiNHLJd5ORy-0BR",
          "type": "internal",
          "value": 1000000000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 13265n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "cell": "x{ED81304A800784F79BB9BDDE39E2B05A2820A591FBCD92F3ACC083685CD65E37ED00489A4B200000000000000C80000000000000003_}",
              "type": "cell",
            },
            "bounce": true,
            "from": "kQDhiRIXBxafSMMm6UXqSB2b7t0IWaAHrHiNHLJd5ORy-0BR",
            "to": "kQCmw0-Nvnap6F8s-77yXEYbduFtd_bSsmDZu6eYT4Oy-Aqh",
            "type": "internal",
            "value": 398473000n,
          },
        ],
      },
    ],
  },
]
`;

exports[`PositionsManagerContract on witdrawStablecoin from positionsManager should revert unhealthy position 1`] = `
[
  {
    "$seq": 3,
    "events": [
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{4A35F9D0800784F79BB9BDDE39E2B05A2820A591FBCD92F3ACC083685CD65E37ED00489A4B20000000000000000000000000000000000000000000000000000000000000017}",
            "type": "cell",
          },
          "bounce": true,
          "from": "kQCmw0-Nvnap6F8s-77yXEYbduFtd_bSsmDZu6eYT4Oy-Aqh",
          "to": "kQDhiRIXBxafSMMm6UXqSB2b7t0IWaAHrHiNHLJd5ORy-0BR",
          "type": "internal",
          "value": 1000000000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 20413n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "cell": "x{5DA6DE96800784F79BB9BDDE39E2B05A2820A591FBCD92F3ACC083685CD65E37ED00489A4B30038624485C1C5A7D230C9BA517A920766FBB742166801EB1E23472C9779391CBEE_}",
              "type": "cell",
            },
            "bounce": true,
            "from": "kQDhiRIXBxafSMMm6UXqSB2b7t0IWaAHrHiNHLJd5ORy-0BR",
            "to": "kQCe9rE3xPdst3hC14pLYI445Rop9zC6414J3Dar2QoLJmVd",
            "type": "internal",
            "value": 966560000n,
          },
        ],
      },
    ],
  },
  {
    "$seq": 4,
    "events": [
      {
        "$type": "received-bounced",
        "message": {
          "body": {
            "type": "empty",
          },
          "bounce": false,
          "from": "kQCe9rE3xPdst3hC14pLYI445Rop9zC6414J3Dar2QoLJmVd",
          "to": "kQDhiRIXBxafSMMm6UXqSB2b7t0IWaAHrHiNHLJd5ORy-0BR",
          "type": "internal",
          "value": 959406000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 7461n,
      },
    ],
  },
]
`;

exports[`PositionsManagerContract on withdraw collateral decrease position collateral  1`] = `
[
  {
    "$seq": 15,
    "events": [
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{CB461808800784F79BB9BDDE39E2B05A2820A591FBCD92F3ACC083685CD65E37ED00489A4B22C8000000020000000200000002000000020000000200000002935F1B401_}",
            "type": "cell",
          },
          "bounce": true,
          "from": "kQCmw0-Nvnap6F8s-77yXEYbduFtd_bSsmDZu6eYT4Oy-Aqh",
          "to": "kQDhiRIXBxafSMMm6UXqSB2b7t0IWaAHrHiNHLJd5ORy-0BR",
          "type": "internal",
          "value": 1000000000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 13714n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "cell": "x{C3CD33BF800784F79BB9BDDE39E2B05A2820A591FBCD92F3ACC083685CD65E37ED00489A4B200000000000000C9_}",
              "type": "cell",
            },
            "bounce": true,
            "from": "kQDhiRIXBxafSMMm6UXqSB2b7t0IWaAHrHiNHLJd5ORy-0BR",
            "to": "kQCmw0-Nvnap6F8s-77yXEYbduFtd_bSsmDZu6eYT4Oy-Aqh",
            "type": "internal",
            "value": 98537000n,
          },
        ],
      },
    ],
  },
]
`;

exports[`PositionsManagerContract on withdraw collateral revert if more than exists  1`] = `
[
  {
    "$seq": 14,
    "events": [
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{CB461808800784F79BB9BDDE39E2B05A2820A591FBCD92F3ACC083685CD65E37ED00489A4B232C000000020000000200000002000000020000000200000002935F1B401_}",
            "type": "cell",
          },
          "bounce": true,
          "from": "kQCmw0-Nvnap6F8s-77yXEYbduFtd_bSsmDZu6eYT4Oy-Aqh",
          "to": "kQDhiRIXBxafSMMm6UXqSB2b7t0IWaAHrHiNHLJd5ORy-0BR",
          "type": "internal",
          "value": 1000000000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 13714n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "cell": "x{C3CD33BF800784F79BB9BDDE39E2B05A2820A591FBCD92F3ACC083685CD65E37ED00489A4B2000000000000012D_}",
              "type": "cell",
            },
            "bounce": true,
            "from": "kQDhiRIXBxafSMMm6UXqSB2b7t0IWaAHrHiNHLJd5ORy-0BR",
            "to": "kQCmw0-Nvnap6F8s-77yXEYbduFtd_bSsmDZu6eYT4Oy-Aqh",
            "type": "internal",
            "value": 98537000n,
          },
        ],
      },
    ],
  },
]
`;

exports[`PositionsManagerContract on withdrawStablecoin from positionsManager with healthy position 1`] = `
[
  {
    "$seq": 7,
    "events": [
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{6C3989BB800784F79BB9BDDE39E2B05A2820A591FBCD92F3ACC083685CD65E37ED00489A4B200000004A817C800000000020000000200000002000000027735940000000002935F1B401_}",
            "type": "cell",
          },
          "bounce": true,
          "from": "kQCmw0-Nvnap6F8s-77yXEYbduFtd_bSsmDZu6eYT4Oy-Aqh",
          "to": "kQDhiRIXBxafSMMm6UXqSB2b7t0IWaAHrHiNHLJd5ORy-0BR",
          "type": "internal",
          "value": 1000000000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 35875n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "cell": "x{903559B2800784F79BB9BDDE39E2B05A2820A591FBCD92F3ACC083685CD65E37ED00489A4B2A04A817C801_}",
              "type": "cell",
            },
            "bounce": true,
            "from": "kQDhiRIXBxafSMMm6UXqSB2b7t0IWaAHrHiNHLJd5ORy-0BR",
            "to": "kQCDr0Gm-nivCD9KWPb6JkiDulGZ8eNb2Xm_HNi39oyNbBwJ",
            "type": "internal",
            "value": 98557000n,
          },
        ],
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "cell": "x{96096519800784F79BB9BDDE39E2B05A2820A591FBCD92F3ACC083685CD65E37ED00489A4B200000004A817C801_}",
              "type": "cell",
            },
            "bounce": true,
            "from": "kQDhiRIXBxafSMMm6UXqSB2b7t0IWaAHrHiNHLJd5ORy-0BR",
            "to": "kQCmw0-Nvnap6F8s-77yXEYbduFtd_bSsmDZu6eYT4Oy-Aqh",
            "type": "internal",
            "value": 962662000n,
          },
        ],
      },
    ],
  },
  {
    "$seq": 8,
    "events": [
      {
        "$type": "received-bounced",
        "message": {
          "body": {
            "type": "empty",
          },
          "bounce": false,
          "from": "kQCDr0Gm-nivCD9KWPb6JkiDulGZ8eNb2Xm_HNi39oyNbBwJ",
          "to": "kQDhiRIXBxafSMMm6UXqSB2b7t0IWaAHrHiNHLJd5ORy-0BR",
          "type": "internal",
          "value": 97557000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 7461n,
      },
    ],
  },
]
`;

exports[`PositionsManagerContract on withdrawStablecoin from positionsManager with unhealthy position 1`] = `
[
  {
    "$seq": 6,
    "events": [
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{6C3989BB800784F79BB9BDDE39E2B05A2820A591FBCD92F3ACC083685CD65E37ED00489A4B2000001D1A94A2000000000020000000200000002000000027735940000000002935F1B401_}",
            "type": "cell",
          },
          "bounce": true,
          "from": "kQCmw0-Nvnap6F8s-77yXEYbduFtd_bSsmDZu6eYT4Oy-Aqh",
          "to": "kQDhiRIXBxafSMMm6UXqSB2b7t0IWaAHrHiNHLJd5ORy-0BR",
          "type": "internal",
          "value": 1000000000n,
        },
      },
      {
        "$type": "failed",
        "errorCode": 43504,
        "errorMessage": "position will not be healthy",
      },
      {
        "$type": "sent-bounced",
        "message": {
          "body": {
            "type": "empty",
          },
          "bounce": false,
          "from": "kQDhiRIXBxafSMMm6UXqSB2b7t0IWaAHrHiNHLJd5ORy-0BR",
          "to": "kQCmw0-Nvnap6F8s-77yXEYbduFtd_bSsmDZu6eYT4Oy-Aqh",
          "type": "internal",
          "value": 974036000n,
        },
      },
    ],
  },
]
`;

exports[`PositionsManagerContract should deploy correctly and send newPositionId request to positionsManager 1`] = `
[
  {
    "$seq": 0,
    "events": [
      {
        "$type": "deploy",
      },
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{946A98B60000000000000000}",
            "type": "cell",
          },
          "bounce": true,
          "from": "kQAI-3FJVc_ywSuY4vq0bYrzR7S4Och4y7bTU_i5yLOB3A6P",
          "to": "kQDhiRIXBxafSMMm6UXqSB2b7t0IWaAHrHiNHLJd5ORy-0BR",
          "type": "internal",
          "value": 1000000000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 16029n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "cell": "x{0D74C595800784F79BB9BDDE39E2B05A2820A591FBCD92F3ACC083685CD65E37ED00489A4B3_}",
              "type": "cell",
            },
            "bounce": true,
            "from": "kQDhiRIXBxafSMMm6UXqSB2b7t0IWaAHrHiNHLJd5ORy-0BR",
            "to": "kQCmw0-Nvnap6F8s-77yXEYbduFtd_bSsmDZu6eYT4Oy-Aqh",
            "type": "internal",
            "value": 98601000n,
          },
        ],
      },
    ],
  },
  {
    "$seq": 1,
    "events": [
      {
        "$type": "received-bounced",
        "message": {
          "body": {
            "type": "empty",
          },
          "bounce": false,
          "from": "kQCmw0-Nvnap6F8s-77yXEYbduFtd_bSsmDZu6eYT4Oy-Aqh",
          "to": "kQDhiRIXBxafSMMm6UXqSB2b7t0IWaAHrHiNHLJd5ORy-0BR",
          "type": "internal",
          "value": 97601000n,
        },
      },
      {
        "$type": "processed",
        "gasUsed": 7461n,
      },
    ],
  },
]
`;
